pkg/api/vm/handler.go

```diff
@@ -515,17 +566,40 @@ func (h *vmActionHandler) removeVolume(ctx context.Context, namespace, name stri
 		Error()
 }
 
+func sanitizeVirtualMachineForTemplate(template *harvesterv1.VirtualMachineTemplate, vm *kv1.VirtualMachine) harvesterv1.VirtualMachineSourceSpec {
+	sanitizedVm := removeMacAddresses(vm)
+	sanitizedVm = replaceCloudInitSecrets(template, sanitizedVm)
+
+	return harvesterv1.VirtualMachineSourceSpec{
+		ObjectMeta: sanitizedVm.ObjectMeta,
+		Spec:       sanitizedVm.Spec,
+	}
+}
+
+func replaceCloudInitSecrets(template *harvesterv1.VirtualMachineTemplate, vm *kv1.VirtualMachine) *kv1.VirtualMachine {
+	sanitizedVm := vm.DeepCopy()
+	for index, volume := range sanitizedVm.Spec.Template.Spec.Volumes {
+		if volume.CloudInitNoCloud == nil {
+			continue
+		}
+		if volume.CloudInitNoCloud.UserDataSecretRef != nil {
+			sanitizedVm.Spec.Template.Spec.Volumes[index].CloudInitNoCloud.UserDataSecretRef.Name = fmt.Sprintf("templateversion-%s-%s-userdata", template.Name, volume.Name)
+		}
+		if volume.CloudInitNoCloud.NetworkDataSecretRef != nil {
+			sanitizedVm.Spec.Template.Spec.Volumes[index].CloudInitNoCloud.NetworkDataSecretRef.Name = fmt.Sprintf("templateversion-%s-%s-networkdata", template.Name, volume.Name)
```

Comment:
this is not identical with the above `createCloudConfigSecrets` method as it uses `template.Name` instead of the `templateVersion.Name`.